app/src/commonMain/kotlin/io/github/cgr/app/io/Unpack.kt

```diff
@@ -0,0 +1,47 @@
+package io.github.cgr.app.io
+
+import io.github.cgr.app.Program
+import io.github.cgr.app.objects.ConstantLabel
+import io.github.cgr.app.objects.Opcode
+
+fun readProgram(buffer: ByteArray): Program {
```

Comment:
Just to make sure I understand this correctly, the binary format is:

- symbol table size: Int
- symbol table: String[]
- constant pool size: Int
- constant pool: Pair<Int, Any?>[] (first value of pair is index in symbol table)
- program length: Int
- opcodes: Int[]  
  opcode format:
   - index of opcode name in symbol table
   - number of args, if variable
   - args, if applicable

right? What happens if the program is long enough that there's more data after the given number of opcodes are read?

Reply:
Yep. The extra data will be ignored

Reply:
How does it know which type to decode each entry in the constant pool as?

Reply:
@Seggan You marked this as resolved, but didn't answer my last question.

Reply:
I did answer
> The extra data will be ignored

> How does it know which type to decode each entry in the constant pool as?
As for that, the `ConstantLabel#getConstant` method takes care of that. Check the `BaseLabel` class for example.

Reply:
So it looks like the entry in the symbol table, which the first element of the pair points to, names the type. Is that correct?

Reply:
Correct